### PR TITLE
[REST] OpenAI Rest API compatible function calling

### PR DESCRIPTION
**Pull request by Kartik14**
This PR adds rest API function calling feature. As per the open AI documentation, the recipe for function api calling is as follows:

> The basic sequence of steps for function calling is as follows:
> 
> - Call the model with the user query and a set of functions defined in the [functions parameter](https://platform.openai.com/docs/api-reference/chat/create#chat/create-functions).
> - The model can choose to call one or more functions; if so, the content will be a stringified JSON object adhering to your custom schema (note: the model may hallucinate parameters).
> - Parse the string into JSON in your code, and call your function with the provided arguments if they exist.
> - Call the model again by appending the function response as a new message, and let the model summarize the results back to the user.

Example:

```python
# test_function_calling
tools = [
    {
        "type": "function",
        "function": {
            "name": "get_current_weather",
            "description": "Get the current weather in a given location",
            "parameters": {
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": "The city and state, e.g. San Francisco, CA",
                    },
                    "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
                },
                "required": ["location"],
            },
        },
    }
]

payload = {
    "model": "gorilla-openfunctions-v1-q4f16_1",
    "messages": [
        {
            "role": "user",
            "content": "What is the current weather in Pittsburgh and tokyo?",
        },
    ],
    "stream": False,
    "temperature": 0.7,
    "tools": tools,
}
r = requests.post("http://127.0.0.1:8000/v1/chat/completions", json=payload)
print(f"Without streaming:\n{r.json()['choices'][0]['message']['tool_calls']}\n")
```

Expected output:
```
[
  {
    "id": "call_y5dfxRxkPvedCUmpfYB7Ur",
    "type": "function",
    "function": {
      "name": "get_current_weather",
      "arguments": {
        "location": "Pittsburgh,PA"
      }
    }
  },
  {
    "id": "call_VFdQdLHUWnw5MewjNFmzzh",
    "type": "function",
    "function": {
      "name": "get_current_weather",
      "arguments": {
        "location": "Tokyo,JP"
      }
    }
  }
]
```

References:
- [OpenAI API reference](https://platform.openai.com/docs/api-reference/chat/create?lang=python)
- [Cookbook examples](https://cookbook.openai.com/examples/how_to_call_functions_with_chat_models)
- #1341 

Comments:

The model `gorilla-openfunctions-v1-q4f16_1` is a bit unstable. For example, changing order of functions leads to different outcomes. Also currently, the function calling assumes that all the api parameters are names parameters which the model does not always respect.

### CODE REVIEW

File: python/mlc_chat/rest.py [hunk 1]
```diff
@@ -180,6 +184,69 @@ async def __anext__(self):
         raise StopAsyncIteration
 
 
+def add_function_call(prompt: List[ChatMessage], function_string: str):
+    # update content of the last input message to include function string
+    user_query = prompt[-1].content
+    prompt[-1].content = f"<<question>> {user_query} <<function>> {function_string}\n"
```
Comment (MasterJH5574) [inline python/mlc_chat/rest.py:190]: Read a bit, looks like the `<<question>>` `<<function>>` here are specific to the gorilla model. This indicates that we may need to extend the conversation protocol to incorporate how user queries interact with functions.

cc @tqchen

**Review by MasterJH5574 — COMMENTED** [review]
Thank you @Kartik14 for the great work! Left a few comments here.

**Comment by MasterJH5574** [comment]
Besides, I noted that the gorilla model uses `USER: `/`ASSISTANT: ` as the conversation template https://github.com/ShishirPatil/gorilla/tree/fcceefe5c5c8beb25ee28f80331d26d8946ad294/openfunctions#running-openfunctions-locally. We need to make sure to use the correct template when compiling this model and uploading the pre-built gorilla models to HF.

File: python/mlc_chat/rest.py (hunk 1 above)
Comment (tqchen) [reply python/mlc_chat/rest.py:190]: let us survey a few other models that support function call, then update template.

File: python/mlc_chat/rest.py [hunk 2]
```diff
@@ -208,6 +275,9 @@ async def request_chat_completion(request: ChatCompletionRequest):
 
     session["chat_mod"].reset_chat()  # Reset previous history, KV cache, etc.
 
+    use_function_call = function_call_util(request)
+    print("Request message after function call util", request.messages)
```
[the remark this answers is not in the document]
Comment (Kartik14) [reply python/mlc_chat/rest.py]: Removed it.

**Comment by t-dettling** [comment]
@Kartik14 This is really great work, thank you for your work on that!

**Comment by t-dettling** [comment]
> Besides, I noted that the gorilla model uses `USER: `/`ASSISTANT: ` as the conversation template https://github.com/ShishirPatil/gorilla/tree/fcceefe5c5c8beb25ee28f80331d26d8946ad294/openfunctions#running-openfunctions-locally. We need to make sure to use the correct template when compiling this model and uploading the pre-built gorilla models to HF.

If I want to compile a model locally, how can I also make sure that I am using the right template? Is there a way that I can specify or edit the template before I compile a model?

**Pull request merged**